src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -1914,6 +1931,37 @@ void ClientChannel::CallData::Destroy(
   }
 }
 
+void ClientChannel::CallData::PreCancel(grpc_call_element* elem,
+                                        grpc_error_handle error) {
+  auto* calld = static_cast<CallData*>(elem->call_data);
+  auto* chand = static_cast<ClientChannel*>(elem->channel_data);
+  // Check if we're queued pending a resolver result.
+  {
+    MutexLock lock(&chand->resolution_mu_);
+    if (calld->queued_pending_resolver_result_) {
+      if (GRPC_TRACE_FLAG_ENABLED(grpc_client_channel_routing_trace)) {
+        gpr_log(GPR_INFO,
+                "chand=%p calld=%p: cancelling resolver queued pick: %s", chand,
+                calld, grpc_error_string(error));
+      }
+      // Remove pick from list of queued picks.
+      calld->MaybeRemoveCallFromResolverQueuedCallsLocked(elem);
+      // Fail pending batches on the call.
+      calld->PendingBatchesFail(elem, error,
+                                YieldCallCombinerIfPendingBatchesFound);
+      return;
+    }
+  }
+  // Not pending resolver result, so check if we have a dynamic call.
+  MutexLock lock(&calld->dynamic_call_creation_mu_);
+  if (calld->dynamic_call_ != nullptr) {
+    calld->dynamic_call_->PreCancel(error);
```

Comment:
I'm not sure how long `PreCancel` would take in general, but does this call have to happen under lock? (e.g., could we just get the pointer and store a flag that the call is being run?)

Reply:
Good point!  Done.